Sources/MongoSwift/BSON/Document.swift

```diff
@@ -11,57 +11,51 @@ internal typealias MutableBSONPointer = UnsafeMutablePointer<bson_t>
 
 /// The storage backing a MongoSwift `Document`.
 public class DocumentStorage {
-    internal var pointer: MutableBSONPointer!
-
-    // Normally, this would go under Document, but computed properties cannot be used before all stored properties are
-    // initialized. Putting this under DocumentStorage gives a correct count and use of it inside of an init() as long
-    // as we have initialized Document.storage beforehand.
-    internal var count: Int {
-        return Int(bson_count_keys(self.pointer))
-    }
+    internal var _bson: MutableBSONPointer!
 
     /// Initializes a new, empty `DocumentStorage`.
     internal init() {
-        self.pointer = bson_new()
+        self._bson = bson_new()
     }
 
     /// Initializes a new `DocumentStorage` by copying over the data from the provided `bson_t`.
     internal init(copying pointer: BSONPointer) {
-        self.pointer = bson_copy(pointer)
+        self._bson = bson_copy(pointer)
     }
 
     /// Initializes a new `DocumentStorage` that uses the provided `bson_t` as its backing storage.
     /// The newly created instance will handle cleaning up the pointer upon deinitialization.
     internal init(stealing pointer: MutableBSONPointer) {
-        self.pointer = pointer
+        self._bson = pointer
     }
 
     /// Cleans up internal state.
     deinit {
-        guard let pointer = self.pointer else {
+        guard let bson = self._bson else {
             return
         }
 
-        bson_destroy(pointer)
-        self.pointer = nil
+        bson_destroy(bson)
+        self._bson = nil
     }
 }
 
 /// A struct representing the BSON document type.
 @dynamicMemberLookup
 public struct Document {
-    /// the storage backing this document
-    internal var storage: DocumentStorage
-
-    /// Returns the number of (key, value) pairs stored at the top level of this `Document`.
-    public internal(set) var count: Int
+    /// the storage backing this document.
+    internal var _storage: DocumentStorage
```

Comment:
Just to clarify: what is the convention around underscore naming / what do they signify now?

Reply:
I read our conversation earlier as that we should use them for anything besides a public property. but I don't feel very strongly so I'm open to any other convention on this

Reply:
I think just internal vs public might be too broad. I think it'd be more useful to designate things we truly want as `fileprivate` or `private`. I know a lot of times I'll update some private field to `internal` so I can use it elsewhere, but doing that is not a great practice and leads to problems like what this PR is trying to solve. We could use the underscore naming as a way to say "keep this thing private".

That said, I too don't feel that strongly about this (and have introduced stuff that doesn't align with the above convention myself), so if that convention doesn't make sense I'm fine with not doing it. I do think that if we're going to name some stuff with underscores, we should settle on some convention around them, though.

Reply:
yeah, without a real convention on what the underscore means, I don't think it gives us any extra information. that seems like a reasonable convention to use, though yeah it definitely doesn't align with a lot of the driver now... lots of our `_x` properties are actually internal. 

alternatively we can get rid of all of the underscores and say this is what swift's fine-grained ACLs are for.

Reply:
Both options are fine with me really.

Reply:
afaict the underscore is used in Apple's code to indicate a fileprivate

Reply:
I don't want to hold up this PR for style nits, so it's not super important what happens to them here. I do think we should either revisit the convention in the future or remove all of them from the codebase if we find that such a convention isn't necessary.

Reply:
I've opened a follow up ticket [SWIFT-498](https://jira.mongodb.org/browse/SWIFT-498)